api/debug.go

```diff
@@ -95,3 +130,241 @@ func (d *DebugAPI) TraceBlockByHash(
+	}
+
+	return tracer.GetResult()
+}
+
+func (d *DebugAPI) executorAtBlock(block *models.Block) (*evm.BlockExecutor, error) {
+	blockHeight := block.Height
+	client, err := d.client.GetClientForHeight(blockHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	exeClient, ok := client.(*grpc.Client)
+	if !ok {
+		return nil, fmt.Errorf("could not convert to execution client")
+	}
+
+	ledger, err := evm.NewRemoteLedger(exeClient.ExecutionDataRPCClient(), blockHeight)
+	if err != nil {
+		return nil, fmt.Errorf("could not create remote ledger for height: %d, with: %w", blockHeight, err)
+	}
+
+	return evm.NewBlockExecutor(
+		block,
+		ledger,
+		d.config.FlowNetworkID,
+		d.blocks,
+		d.receipts,
+		d.logger,
+	)
+}
+
+func applyStateOverrides(config *tracers.TraceCallConfig, stateDB types.StateDB) error {
+	if config == nil || config.StateOverrides == nil {
+		return nil
+	}
+
+	diff := *config.StateOverrides
+	for addr, account := range diff {
+		// Override account nonce.
+		if account.Nonce != nil {
+			stateDB.SetNonce(addr, uint64(*account.Nonce))
+		}
+		// Override account(contract) code.
+		if account.Code != nil {
+			stateDB.SetCode(addr, *account.Code)
+		}
+		// Override account balance.
+		if account.Balance != nil {
+			u256Balance, _ := uint256.FromBig((*big.Int)(*account.Balance))
+			currentBalance := stateDB.GetBalance(addr)
+			// If given balance is greater-or-equal than the current balance,
+			// we need to add the diff to the current balance. Otherwise,
+			// we need to sub the diff from the current balance.
+			if u256Balance.Cmp(currentBalance) >= 0 {
+				diff := u256Balance.Sub(u256Balance, currentBalance)
+				stateDB.AddBalance(addr, diff, tracing.BalanceChangeUnspecified)
+			} else {
+				diff := currentBalance.Sub(currentBalance, u256Balance)
+				stateDB.SubBalance(addr, diff, tracing.BalanceChangeUnspecified)
+			}
+		}
+		if account.State != nil && account.StateDiff != nil {
+			return fmt.Errorf("account %s has both 'state' and 'stateDiff'", addr.Hex())
+		}
+		// Replace entire state if caller requires.
+		if account.State != nil {
+			for key, value := range *account.State {
+				stateDB.SetState(addr, key, value)
+			}
+		}
+		// Apply state diff into specified accounts.
+		if account.StateDiff != nil {
+			for key, value := range *account.StateDiff {
+				stateDB.SetState(addr, key, value)
+			}
+		}
+	}
+
+	return stateDB.Commit(true)
+}
+
+func (d *DebugAPI) resolveBlockNumberOrHash(block *rpc.BlockNumberOrHash) (uint64, error) {
+	err := fmt.Errorf("%w: neither block number nor hash specified", errs.ErrInvalid)
+	if block == nil {
+		return 0, err
+	}
+	if number, ok := block.Number(); ok {
+		return d.resolveBlockNumber(number)
+	}
+
+	if hash, ok := block.Hash(); ok {
+		evmHeight, err := d.blocks.GetHeightByID(hash)
+		if err != nil {
+			return 0, err
+		}
+		return evmHeight, nil
+	}
+
+	return 0, err
+}
+
+func (d *DebugAPI) resolveBlockNumber(number rpc.BlockNumber) (uint64, error) {
```

Comment:
there is a similar function in `services/requester/requester.go` can we make a general function for it?

Reply:
I was not able to find anything similar in the `services/requester/requester.go`. Maybe it was in a different PR?